src/Features/Core/Portable/IncrementalCaches/SymbolTreeInfoIncrementalAnalyzerProvider.cs

```diff
@@ -1,100 +1,215 @@
+                return Task.FromResult(default(ValueTuple<bool, SymbolTreeInfo>));
+            }
 
-                var info = SymbolTreeInfo.Create(VersionStamp.Default, compilation.Assembly, cancellationToken);
-                if (info != null)
+            public async Task<ValueTuple<bool, SymbolTreeInfo>> TryGetSymbolTreeInfoAsync(Project project, CancellationToken cancellationToken)
+            {
+                Tuple<VersionStamp, SymbolTreeInfo> tuple;
+                if (_projectToInfo.TryGetValue(project.Id, out tuple))
                 {
-                    RecordCount(project.Id, info.Count);
+                    var version = await project.GetSemanticVersionAsync(cancellationToken).ConfigureAwait(false);
+                    if (version == tuple.Item1)
+                    {
+                        return ValueTuple.Create(true, tuple.Item2);
+                    }
                 }
+
+                return default(ValueTuple<bool, SymbolTreeInfo>);
+            }
+        }
+
+        private class IncrementalAnalyzer : IncrementalAnalyzerBase
+        {
+            private readonly ConcurrentDictionary<ProjectId, Tuple<VersionStamp, SymbolTreeInfo>> _projectToInfo;
+
+            // Note: the Incremental-Analyzer infrastructure guarantees that it will call all the methods
+            // on this type in a serial fashion.  As such, we don't need explicit locking, or threadsafe
+            // collections (if they're only used by this type).  So, for example, the map we populate
+            // needs to be a ConcurrentDictionary as it will be read and written from multiple types.
+            // However, the HashSet<ProjectId> is ok as it will only be used by this type and there is
+            // no concurrency in this type on its own.
+            private readonly ConcurrentDictionary<string, Tuple<DateTime, SymbolTreeInfo, HashSet<ProjectId>>> _metadataPathToInfo;
+
+            public IncrementalAnalyzer(
+                ConcurrentDictionary<ProjectId, Tuple<VersionStamp, SymbolTreeInfo>> projectToInfo,
+                ConcurrentDictionary<string, Tuple<DateTime, SymbolTreeInfo, HashSet<ProjectId>>> metadataPathToInfo)
+            {
+                _projectToInfo = projectToInfo;
+                _metadataPathToInfo = metadataPathToInfo;
             }
 
-            public override Task NewSolutionSnapshotAsync(Solution solution, CancellationToken cancellationToken)
+            public override async Task AnalyzeProjectAsync(Project project, bool semanticsChanged, CancellationToken cancellationToken)
             {
-                // check whether we are good to report total symbol numbers
-                if (_symbolCountByProjectMap == null || _symbolCountByProjectMap.Count < solution.ProjectIds.Count || string.IsNullOrEmpty(solution.FilePath))
+                if (!project.SupportsCompilation || project.FilePath == null)
```

Comment:
Yeah, I don't know why I did that.  I think it was because I was originally planning on using the filepath.  but that didn't end up happening, so I can def remove this check now.